File: javascript/aho_corasick_string_search/aho_corasick_node.js
const AhoCorasickNode = function (letter) {
  this.word = null;
  this.letter = letter;
  this.prevLink = null;
  this.outputLink = null;
  this.suffixLink = null;
  this.children = {};
};

AhoCorasickNode.prototype.getChildOfLetter = function (letter) {
  return this.children[letter];
};

AhoCorasickNode.prototype.addChild = function (letter) {
  const newNode = new AhoCorasickNode(letter);
  newNode.prevLink = this;
  this.children[letter] = newNode;
  return newNode;
};

AhoCorasickNode.prototype.setWord = function (word) {
  this.word = word;
};

AhoCorasickNode.prototype.setOutputLink = function (outputLinkNode) {
  this.outputLink = outputLinkNode;
};

AhoCorasickNode.prototype.setSuffixLink = function (suffixLinkNode) {
  this.suffixLink = suffixLinkNode;
};

AhoCorasickNode.recursivePrint = function (start, depth) {
  const realDepth = depth || 0;

  let outputString = '';
  for (let i = 0; i < realDepth; i++) {
    outputString += "----|";
  }
  outputString += " " + start.letter;
  if (start.word) {
    outputString += " (" + start.word + ")";
  }
  if (start.suffixLink) {
    outputString += " [" + start.suffixLink.letter + "]";
  }
  if (start.outputLink) {
    outputString += " {" + start.outputLink.word + "}";
  }
  console.log(outputString);

  const childKeys = Object.keys(start.children);
  for (let key of childKeys) {
    const child = start.children[key];
    AhoCorasickNode.recursivePrint(child, realDepth + 1);
  }
};

AhoCorasickNode.prototype.getAllChildren = function () {
  return Object.values(this.children);
};

module.exports = AhoCorasickNode;

File: ruby/aho_corasick_string_search/aho_corasick_node.rb
class AhoCorasickNode

  attr_reader :letter, :children
  attr_accessor :word, :letters_sum, :prev_link, :suffix_link, :output_link

  def initialize(letter)
    @letter = letter
    @children = []
    @word = nil
    @prev_link = nil
    @suffix_link = nil
    @output_link = nil
    @letters_sum = nil
  end

  def child_of_letter(l)
    @children.find { |e| e.letter == l  }
  end

  def add_child(child_node)
    @children << child_node
  end

  def display(depth = 0)
    depth.times { print '----|' }
    print " #{@letter} "
    if @word
      print "( #{@word} )"
    end
    if @suffix_link
      print "[ #{@suffix_link.letter} / #{@suffix_link.letters_sum} ]"
    end
    if @output_link
      print "{ #{@output_link.word} }"
    end
    puts

    children.each { |c| c.display(depth + 1) }
  end

end
